# OS Walkthrough

In [3]:
!pip install opensearch-py

## Libraries Used

In [160]:
!pip install uuid

Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://semple_devs:****@nexus.global.lmco.com/repository/pypi-proxy/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for uuid: filename=uuid-1.30-py3-none-any.whl size=6501 sha256=3125612e403da73b4a1fe79c81c622b869df7cc4f97ce4ee910870229369b86a
  Stored in directory: /home/jovyan/.cache/pip/wheels/60/d9/c6/6e7909dfc39c636e7d25bec82a97d96cc82cd7d7462028e426
Successfully built uuid

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


In [162]:
from opensearchpy import OpenSearch

from datetime import datetime
import random
import string
import copy
from uuid import uuid4
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# AI
from openai import OpenAI
from httpx import Client

ai_client = OpenAI(
    base_url="https://api.ai.us.lmco.com/v1/",
    api_key=os.getenv("AZURE_API_KEY")
    )

## What is OpenSearch?

OpenSearch is an **open-source, distributed search and analytics engine**. It enables developers, data scientists, and analysts to search, index, and analyze large volumes of data in near real-time. It's especially useful for full-text search, log analytics, and dashboarding.

- It supports full-text, fuzzy, and structured search.
- It scales horizontally across multiple machines.
- It integrates with Kibana's fork: **OpenSearch Dashboards**.
- It is open-source and community-driven, maintained by the OpenSearch Project (led by AWS).

You can learn more about the project at the official OpenSearch website: [https://opensearch.org](https://opensearch.org)

Here are a few great beginner-friendly videos to get you up to speed quickly:

1. **What is OpenSearch? (Official Intro by OpenSearch Project)**  
   https://www.youtube.com/watch?v=HOKBxcLHZ5A  
   A 3-minute official overview from the OpenSearch team.

2. **OpenSearch vs Elasticsearch: What's the Difference?**  
   https://www.youtube.com/watch?v=wNBvF0yUUvY  
   A helpful comparison for folks familiar with Elasticsearch.

3. **Search 101: How Search Engines Work (great conceptual foundation)**  
   https://www.youtube.com/watch?v=tj7lVdkS3Dg  
   This one is not OpenSearch-specific but gives great context on how search indexing and queries work in general.

OpenSearch is like a powerful search engine and analytics toolkit, built to help you quickly find and analyze data stored in JSON format.

In the next section, we’ll learn how to **install OpenSearch**, connect to it using Python, and send our first query!


## Trying OS on Local

We’ll use the official OpenSearch Docker images to run a single-node cluster locally.

This command pulls and runs OpenSearch in a Docker container with minimal configuration:

```bash
docker run -d --name opensearch \
  -p 9200:9200 -p 9600:9600 \
  -e "discovery.type=single-node" \
  -e "plugins.security.disabled=true" \
  opensearchproject/opensearch:2.13.0

## You will need to run this in your local to make sure you don't run into permission errors

```bash
!sudo usermod -aG docker $USER

### Docker SandBox

In [3]:
!docker run -d --name opensearch \
  -p 9200:9200 -p 9600:9600 \
  -e "discovery.type=single-node" \
  -e "plugins.security.disabled=true" \
  -e "OPENSEARCH_INITIAL_ADMIN_PASSWORD=StrongP@ssw0rd!" \
  opensearchproject/opensearch:2.13.0

docker: Error response from daemon: Conflict. The container name "/opensearch" is already in use by container "45fbdcc4afecd786860d74cc291412aca355c1b4fbaceb1c5d1624b48fabbfd6". You have to remove (or rename) that container to be able to reuse that name.

Run 'docker run --help' for more information


In [7]:
!docker ps

CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                          NAMES
1904351271d0   opensearchproject/opensearch:2.13.0   "./opensearch-docker…"   24 seconds ago   Up 23 seconds   0.0.0.0:9200->9200/tcp, [::]:9200->9200/tcp, 9300/tcp, 0.0.0.0:9600->9600/tcp, [::]:9600->9600/tcp, 9650/tcp   opensearch


### Check to see if it's running

In [1]:
!curl -X GET "localhost:9200"

We **WON'T** be using the local version of OS for the rest of this demo, but I wanted to give it to you in case you want to explore a sandboxed version.

## Our Version

Giving you the password like this isn't great, but I have a feeling that no one has given you the contact information. "With great power, comes great....yada yada"

In [250]:
os.environ['no_proxy'] += ',' + ESHOST

# Pulling summarized docs from OpenSearch
from opensearchpy import OpenSearch, Urllib3HttpConnection, Urllib3AWSV4SignerAuth

client = OpenSearch(
    hosts = [{'host': ESHOST, 'port': 443}],
    http_auth = (ESUSER, ESPASSWORD),
    use_ssl = True,
    verify_certs = True,
)

client.info()

{'name': 'd0cc6056ffaed074b535425f8dcdfd33',
 'cluster_name': '181596953521:lmx-testing-2',
 'cluster_uuid': 'mivq_pOBR2GFTMJYdYELdg',
 'version': {'distribution': 'opensearch',
  'number': '2.17.0',
  'build_type': 'tar',
  'build_hash': 'unknown',
  'build_date': '2025-02-14T09:39:30.748828589Z',
  'build_snapshot': False,
  'lucene_version': '9.11.1',
  'minimum_wire_compatibility_version': '7.10.0',
  'minimum_index_compatibility_version': '7.0.0'},
 'tagline': 'The OpenSearch Project: https://opensearch.org/'}

## Understanding `client.info()` Output

When you call `client.info()` in OpenSearch, you're asking the server for basic metadata about itself — like version, identity, and configuration.

Here’s a breakdown of what each field means:

### Top-Level Fields

- **`name`**:  
  This is the name of the OpenSearch node that responded to your request.  
  In this case: `"50ce1a654b3c74459fce9d1ab68c8706"` — probably auto-generated.

- **`cluster_name`**:  
  The name of the entire OpenSearch cluster.  
  Here, it's `"181596953521:lmx-testing-2"` — often includes account or environment info.

- **`cluster_uuid`**:  
  A unique identifier for the cluster — useful for debugging or verifying consistency.


### `version` Block

This section tells you which version of OpenSearch is running, along with build details.

- **`distribution`**:  
  Confirms that it's an official `opensearch` distribution (not Elasticsearch or a custom fork).

- **`number`**:  
  This is the actual version of OpenSearch.  
  You're running version `2.17.0`.

- **`lucene_version`**:  
  OpenSearch is built on top of Apache Lucene, a low-level search library.  
  Lucene version here is `9.11.1`.

- **`build_type`, `build_hash`, `build_date`**:  
  Mostly useful for internal tracking or debugging (you can skip these for now).

- **`minimum_wire_compatibility_version` / `minimum_index_compatibility_version`**:  
  These show compatibility boundaries — for example, OpenSearch 2.17 can still talk to nodes as old as version 7.10.0.


### `tagline`

Just a nice reminder that you're using the OpenSearch project!  
`"The OpenSearch Project: https://opensearch.org/"`

### TL;DR

You’ve connected to a **live OpenSearch node** in a **cluster named `lmx-testing-2`**, running **version 2.17.0**. Everything looks healthy — now you're ready to create an index, add documents, and start querying.

## Indicies: How OS Manages Data


An **index** in OpenSearch is a logical namespace for storing and organizing documents. If you're familiar with relational databases, you can think of an index as roughly analogous to a table.

Each index contains a collection of **documents**, and each document is a structured JSON object.

Indices allow OpenSearch to organize, search, and analyze data efficiently.


### Why Indices Matter



OpenSearch uses indices to:

- Group related documents (e.g., logs, users, transactions)
- Apply specific settings for performance and scaling (shards, replicas, analyzers)
- Define mappings (similar to schemas) that declare how fields should be interpreted
- Apply lifecycle rules (e.g., how long data is retained or when it should be archived)


### Structure of an Index

Each index contains:

- **Documents**: JSON records of structured or semi-structured data.
- **Mappings**: The schema that defines the types of fields (e.g., text, keyword, date, integer).
- **Settings**: Shard and replica configurations, analyzers, and more.

You can create an index explicitly using the OpenSearch Python client. Here's an example that creates an index called `articles` with defined field types:

In [19]:
index_name = "demo"

def test_build():
    result = client.indices.create(
        index=index_name,
        body={
            "settings": {
                "number_of_shards": 2,
                "number_of_replicas": 2
            },
            "mappings": {
                "properties": {
                    "title": {"type": "text"},
                    "author": {"type": "keyword"},
                    "published": {"type": "date"},
                    "content": {"type": "text"}
                }
            }
        }
    )
    return result
    
test_build()

{'acknowledged': True, 'shards_acknowledged': True, 'index': 'demo'}

## Shards and Clusters in OpenSearch

### What is a Cluster?

An OpenSearch **cluster** is a group of one or more nodes (servers) that together hold your data and provide search and indexing capabilities.

- A single-node setup (like most local installations) is still considered a cluster.
- In a production system, multiple nodes work together for fault tolerance and scalability.

Each cluster has:
- A **unique name** (`cluster_name`)
- A **set of nodes**
- One **master node** (can be elected dynamically)
- Optional **data nodes**, **ingest nodes**, and **coordinating nodes**


### What is a Shard?

A **shard** is a horizontal partition of an index. Every index is split into one or more shards.

- Each shard is a fully functional and independent "mini-index."
- Shards allow OpenSearch to distribute data and load across multiple nodes.

There are two types of shards:

- **Primary shard**: Required for storing data. Every document belongs to exactly one primary shard.
- **Replica shard**: A copy of a primary shard. Provides redundancy and improves query throughput.

When you create an index, you can define:

Anyway, if you build an index that already exists, it will throw an error

In [20]:
try: 
    test_build()
except Exception as e:
    print(e)

RequestError(400, 'resource_already_exists_exception', 'index [demo/YmFSDmmJR82HwxvFqP6QEQ] already exists')


Here is how you delete an index....be careful with this one!

In [21]:
# client.indices.delete(index=index_name, ignore=[400,404])

{'acknowledged': True}

## Find what we have now in OS

In [228]:
response = client.indices.get_alias(index="*")
[x for x in response.keys() if x[0] != "."]

['dev-model-description-index-test',
 'ai_string_tests',
 'edited_doc_36bu7cj4',
 'ai_string_tests_prod',
 'portfolio',
 'flattened_sheets',
 'dev-model-description-index']

This is what has already been made...I've been populating ai_string_tests

In [229]:
mapping = client.indices.get_mapping(index="ai_string_tests")
mapping

{'ai_string_tests': {'mappings': {'properties': {'ai_json': {'type': 'nested',
     'properties': {'0': {'properties': {'Detailed Steps': {'type': 'text',
         'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},
        'Step Description': {'type': 'text',
         'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},
        'Step Detailed Description': {'type': 'text',
         'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},
        'Step Expected Result': {'type': 'text',
         'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},
        'Step Number': {'type': 'long'}}},
      '1': {'properties': {'Detailed Steps': {'type': 'text',
         'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},
        'Step Description': {'type': 'text',
         'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},
        'Step Detailed Description': {'type': 'text',
         'fields': {'keyword': {'type': 'ke

### Inspecting an Existing Index: `ai_string_tests`

We can use the OpenSearch client to inspect the structure of an index. There are two main components:

1. **Mappings** – This tells us the field names and their data types.
2. **Settings** – This includes performance-related configurations like the number of shards and replicas.

## Understanding the Mapping for `ai_string_tests`

The mapping for this index defines the structure of documents, including the field names, types, and any additional configurations. Let’s break down each field:

### 1. `ai_json` (type: nested)
- **Type:** `nested`
- **Meaning:** This field is expected to hold an array of objects. Using a nested type allows you to perform complex queries on these objects as if they were separate documents, while still keeping them attached to their parent document. **Essentailly, this is our AI generated steps that were generated from the string**

### 2. `ai_text` (type: text with a sub-field)
- **Type:** `text`
- **Sub-field:** `keyword`
  - **Type:** `keyword`
  - **Setting:** `ignore_above` is set to 200000.
- **Meaning:** The primary field (`ai_text`) is analyzed, which means it is processed (e.g., tokenized) to support full-text search. The `keyword` sub-field stores the same content in a non-analyzed form, useful for exact matches or aggregations. The `ignore_above` parameter ensures that strings longer than 200,000 characters are not indexed in the keyword sub-field. The AI also critiques the sheets, that might be useful for us in the future

### 3. `created` (type: date)
- **Type:** `date`
- **Format:** `yyyy-MM-dd`
- **Meaning:** This field stores dates and is formatted as a four-digit year, two-digit month, and two-digit day (e.g., 2025-04-01). It’s typically used to record when a document was first created.

### 4. `creator` (type: text)
- **Type:** `text`
- **Meaning:** Stores the name or identifier of the person or process that created the document. Since it’s a text field, it will be analyzed for full-text search.

### 5. `extension` (type: text)
- **Type:** `text`
- **Meaning:** Store file extensions or similar metadata. Being a text field, it’s suitable for search and analysis.

### 6. `information` (type: text)
- **Type:** `text`
- **Meaning:** Holds additional descriptive information. As a text field, it supports full-text search operations.

### 7. `keywords` (type: text)
- **Type:** `text`
- **Meaning:** Contains keywords related to the document. This field helps in categorizing or retrieving documents based on associated keywords.

### 8. `last_modified_by` (type: text)
- **Type:** `text`
- **Meaning:** Indicates the person or process that last modified the document. This is useful for tracking changes and accountability.

### 9. `modified` (type: date)
- **Type:** `date`
- **Format:** `yyyy-MM-dd`
- **Meaning:** Records the date when the document was last modified, using the same date format as the `created` field.

### 10. `name` (type: text)
- **Type:** `text`
- **Meaning:** Could be used to store a title or identifying name for the document. As a text field, it is analyzed to support search.

### 11. `other` (type: text)
- **Type:** `text`
- **Meaning:** A general-purpose field for any additional text data that doesn’t fit into the other defined fields.

### 12. `path` (type: text)
- **Type:** `text`
- **Meaning:** Stores original sharpoint  paths. The text type allows for search, though exact matching might require a keyword sub-field if needed.

### 13. `sheet_name` (type: text)
- **Type:** `text`
- **Meaning:** Likely stores the name of a sheet from a spreadsheet or similar document structure.

### 14. `step_json` (type: nested)
- **Type:** `nested`
- **Meaning:** Similar to `ai_json`, this field is designed to store an array of objects. The nested type allows for querying sub-objects independently while keeping them tied to their parent.

### 15. `title` (type: text)
- **Type:** `text`
- **Meaning:** Intended to store the title of the document. It is analyzed for full-text search so that users can search for documents by their title content.

---

### Summary

This mapping defines a flexible structure for the `ai_string_tests` index. It uses:
- **`text` fields** for content that will be analyzed (and in one case, with a `keyword` sub-field for exact matching),
- **`date` fields** for time-related data formatted as `yyyy-MM-dd`, and
- **`nested` fields** to handle arrays of objects in a structured way.

By understanding these field types and configurations, you can design queries that precisely target the data you need and ensure your documents are stored in an efficient, searchable manner.


Honestly, I am not sure if this a perfect mapping and once you get familar with everything, we can tag-team how to make this better. But for now, it's what we got.

## Sample on the Index

In [230]:
query = {
    "query": {
        "match_all": {}
    },
    "size": 100  # Number of documents to return in the sample
}

results = client.search(index="ai_string_tests", body=query)

In [231]:
results["hits"]["hits"][44]['_source']

{'step_json': {'0': {'Step Number': 1,
   'Step Description': 'Launch and search for Create Purchase Requisition Advance App',
   'Step Expected Result': ' '},
  '1': {'Step Number': 2,
   'Step Description': 'Click on the App ',
   'Step Expected Result': 'Create Purchase Requisition Screen displayed'},
  '2': {'Step Number': 3,
   'Step Description': 'Select the appropriate doct type ',
   'Step Expected Result': ' '},
  '3': {'Step Number': 4,
   'Step Description': 'At the Item overview level, insert:\nInsert the relevant "Account Assignment"\nSelect "Item Category"\nInsert the Item\nQuantity\nPlant\nPurchase Group and press Enter\n',
   'Step Expected Result': 'Fields successfully filled'},
  '4': {'Step Number': 5,
   'Step Description': 'At Item details Level, click on:\n"Service" Tab to maintain service reuqirements\n"Acount Assignment" tab to insert relevant Cost elements\n"Enhaced Limits/Limits" tab to insert spend range\nSource of supply to insert appropriate source\n\n\n',


## Interpreting the OpenSearch Search Response

When you execute a search query in OpenSearch using the Python client, the response is a dictionary that contains multiple layers of metadata and actual results. Here’s how to interpret it:

### Top-Level Fields

- **`took`**:  
  Indicates how many milliseconds the query took to execute on the server.  
  This can be used to measure search performance.

- **`timed_out`**:  
  Boolean flag — `False` means the query completed successfully within the timeout window.

- **`_shards`**:  
  Details about how the query was distributed across the shards.
  - `total`: The number of shards queried (based on index configuration).
  - `successful`: How many shards returned results successfully.
  - `failed`: Number of shards that failed to respond (ideally 0).
  - `skipped`: Number of shards skipped due to routing or filtering.

---

### `hits` Object

This section contains the actual search results.

- **`total`**:  
  The total number of documents that matched the query.
  - `value`: The number of matched documents (e.g., 653).
  - `relation`: Can be `eq` (exact) or `gte` (estimated lower bound).

- **`max_score`**:  
  The highest `_score` of all returned hits. Useful for relevance ranking.

- **`hits`**:  
  A list of documents (usually truncated to the top `size` results).

---

### Each Hit in the `hits` List

Each item in the list is a matching document with additional metadata:

- **`_index`**:  
  The name of the index the document belongs to (`ai_string_tests` in this case).

- **`_id`**:  
  The unique document ID assigned by OpenSearch.

- **`_score`**:  
  A relevance score (used mostly in `match` or `multi_match` queries).  
  A score of 1.0 is common in `match_all` or `term` queries with no ranking.

- **`_source`**:  
  The actual content of the document as it was indexed. This includes all the fields such as:
  - `step_json` and `ai_json`: Nested JSON containing structured process steps.
  - `created`, `modified`: Dates of creation and last modification.
  - `name`, `path`, `sheet_name`: File-related metadata.
  - `ai_text`: A large blob of text with critique and reformatted JSON.
  - `title`, `creator`, `keywords`: Metadata for sorting, filtering, or searching.

---

### Summary

This structure allows OpenSearch to return both metadata (for performance monitoring and routing) and actual content (for your application logic or frontend display).

You can access just the documents using:

In [232]:
results = client.search(index="ai_string_tests", body={"query": {"match_all": {}}, "size": 20})
docs = [hit["_source"] for hit in results["hits"]["hits"]]

In [236]:
# This is the 2nd item on the list
docs[1]

{'step_json': {'0': {'Step Number': 1,
   'Step Description': 'Log into the S4 environment being used for this test cycle and run the program to load the projects and wbs structures to be converted  (PLP-90069-C)',
   'Step Expected Result': 'Projects and WBS Structure were loaded successfully; capture the number of WBS Elements that are in the file to load and the number that were successfully loaded; capture any errors to be reported to the data validator so that any issues can be corrected prior to the next load.  If desired, issues can be corrected within the current file and reloaded to improve the load percentage.'},
  '1': {'Step Number': 2,
   'Step Description': 'Log into the S4 environment being used for this test cycle',
   'Step Expected Result': 'Log in successful'},
  '2': {'Step Number': 3,
   'Step Description': 'Validate the Project and WBS Structure within S4 to ensure they have been loaded correctly',
   'Step Expected Result': 'Define, document, and/or update the va

## Exact Search on Path String

The `match_phrase` query looks for the *exact sequence of terms* in the specified field.
It is different from a regular `match`, which allows term reordering and partial matches.
This query will return documents where the `path` field contains the exact phrase "itc-2",
with terms in that specific order.

In [237]:
query = {
  "query": {
    "match_phrase": {
      "path": "itc-3"
    }
  }, "size": 10_000
}
results = client.search(index="ai_string_tests", body=query)

In [239]:
results = client.search(index="ai_string_tests", body=query)
print("Random 10 Documents with exact phrase 'itc-3' in the path field:")
random.shuffle(results["hits"]["hits"])
for num, hit in enumerate(results["hits"]["hits"]):
    print(f'{num:01}\t{hit["_source"]["path"][0:75]:5}...')

Random 10 Documents with exact phrase 'itc-3' in the path field:
0	Data/3.  Integration Test Repository/ITC-3/Asset Mgmt/INT_ATR_009C - Asset ...
1	Data/3.  Integration Test Repository/ITC-3/Finance/WIP/INT_RTR_020 - Indire...
2	Data/3.  Integration Test Repository/ITC-3/Asset Mgmt/INT_ATR_008A - Equipm...
3	Data/3.  Integration Test Repository/ITC-3/Finance/WIP/INT_RTR_020 - Indire...


### Dot notation example for querying inside a nested object


If your mapping contains a structured JSON object like `step_json` or `ai_json`,
you can access subfields using dot notation.
For example: step_json.0.Step Description
Note: This assumes step_json is NOT mapped as a nested type. If it is, you must use a nested query.

### Fuzzy search inside a JSON string field using dot notation example
and an exact phrase match on the 'path' field.

In [242]:
# This query performs:
# - A fuzzy match inside a nested field (ai_json)
# - An exact match_phrase on the 'path' field

nested_fuzzy_query = {
    "query": {
        "bool": {
            "must": [
                {
                    "nested": {
                        "path": "ai_json",
                        "query": {
                            "match": {
                                "ai_json.0.Step Description": {
                                    "query": "login sap hana",
                                    "fuzziness": "AUTO"
                                }
                            }
                        }
                    }
                },
                {
                    "match_phrase": {
                        "path": "3"
                    }
                }
            ]
        }
    }
}

response = client.search(index="ai_string_tests", body=nested_fuzzy_query)

print("Results for fuzzy match inside ai_json (nested) and exact phrase 'itc-2' in path:")
for hit in response["hits"]["hits"]:
    print("Path:", hit["_source"].get("path"))
    print("Step 0 Description:", hit["_source"].get("ai_json", {}).get("0", {}).get("Step Description", ""))
    print("---")

Results for fuzzy match inside ai_json (nested) and exact phrase 'itc-2' in path:
Path: Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_005 - Customer Contract Creation through Customer Billing and Apply Cash (Deliverable Material, Fixed Price, Progress Payment)/01.TC_O2C_Initiate and Monitor Contract_001 - O2C.xlsx
Step 0 Description: Open SAP Logon Pad and select the appropriate S4 HANA system
---
Path: Data/3.  Integration Test Repository/ITC-2/Prod Ops/INT_PTS_008B - DD250 - Customer Sell-off - Source Destination/01. TC_O2C_Initiate and Monitor Contract_001 - PTS.xlsx
Step 0 Description: Open SAP S4 HANA login page via Fiori.
---
Path: Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_005 - Customer Contract Creation through Customer Billing and Apply Cash (Deliverable Material, Fixed Price, Progress Payment)/01.TC_O2C_Initiate and Monitor Contract_001 - O2C.xlsx
Step 0 Description: Open the SAP Fiori login page.
---
Path: Data/3.  Integr

## === WORKFLOW: USER-DRIVEN JSON EDITING IN OPENSEARCH ===

This script models the process your app will perform:
1. Search for a document by exact match (e.g., title)
2. Extract and copy the document for editing
3. Modify the step_json field (simulate user edits)
4. Add a note for auditing (e.g., edit timestamp)
5. Insert the edited document into a new index
I don't think this is a perfect mirror of what we are hoping to do, but we can talk through it. Or you can 

### Step 1: Perform an exact match search on the title field (use the .keyword subfield for exact matching)


In [243]:
search_query = {
    "query": {
        "match_phrase": {
            "path": "3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_001 - Customer Contracts through Customer Billing and Apply Cash (Deliverable Material, Fixed Price, No Financing)/01.TC_O2C_Initiate and Monitor Contract_001 - O2C.xlsx"  }
    }
}

search_result = client.search(index="ai_string_tests", body=search_query)

In [244]:
search_result["hits"]

{'total': {'value': 8, 'relation': 'eq'},
 'max_score': 36.703705,
 'hits': [{'_index': 'ai_string_tests',
   '_id': '104',
   '_score': 36.703705,
   '_source': {'step_json': {'0': {'Step Number': 1,
      'Step Description': 'S4 Analyst to log into SAP S4 HANA via Fiori',
      'Step Expected Result': 'Log in Successful'},
     '1': {'Step Number': 2,
      'Step Description': 'Navigate to the SAP Contract Screen - Manage sales contract ',
      'Step Expected Result': 'Manage sales contract is displayed'},
     '2': {'Step Number': 3,
      'Step Description': 'Search the created Master contract (ZGK) ',
      'Step Expected Result': 'Master contract is displayed'},
     '3': {'Step Number': 4,
      'Step Description': 'Validate all relevant fields information in the Master Contract (ZGK) ',
      'Step Expected Result': 'Header/Line item fields are populated correctly'},
     '4': {'Step Number': 5,
      'Step Description': 'On the master contract screen under tab " Lowerlevelcon

### Step 2 and 3: Extract and Modify matching document

In [249]:
if not search_result["hits"]["hits"]:
    print("No document found with that title.")
else:
    original_doc = search_result["hits"]["hits"][0]["_source"]
    print("Original document retrieved.")

    # Step 3: Deep copy and modify the JSON
    original_id = search_result["hits"]["hits"][0]["_id"]
    modified_doc = copy.deepcopy(original_doc)
    
    # Simulate user editing step_json (you can replace this with UI-driven input later)
    if "step_json" in modified_doc:
        modified_doc["step_json"]["99"] = {
            "Step Number": 99,
            "Step Description": "User-inserted step from editor",
            "Step Expected Result": "Custom logic executed"}

    # Option A: Overwrite existing document by reusing its _id
    overwrite = False  # Set to True if you want to replace the original

    if overwrite:
        response = client.index(index="ai_string_tests", id=original_id, body=modified_doc)
        print(f"Document overwritten (same ID): {original_id}")
    else:
        new_id = str(uuid4())  # Generate a unique ID
        response = client.index(index="ai_string_tests", id=new_id, body=modified_doc)
        print(f"Modified document inserted with new ID: {new_id}")

Original document retrieved.


RequestError: RequestError(400, 'illegal_argument_exception', 'Limit of total fields [1000] has been exceeded')